### PR TITLE
feat: propagate authentication, client request and key headers between operations

### PR DESCRIPTION
**Pull request by fiam**
- Add authentication middleware to internal operations
- Forward Authentication and X-Request-ID headers between operations
- Decode the client request correctly before passing it to function handlers
- Fix returning falsy data from TS operations
- Add tests for all these

Fixes ENG-2216
Fixes ENG-2215
Fixes ENG-1383
Fixes ENG-1179
Fixes ENG-1073

## Motivation and Context

Internal operations didn't support these features that external operations did.
Additionally, some features didn't work properly when they involved multiple
operation calls.

## Marketing Changelog

Authentication, request logging and accesing client headers now work with internal
operations.

#### Checklist

- [ ] run `make test`
- [ ] PR title must follow [conventional-commit-standard](https://github.com/wundergraph/wundergraph/blob/main/CONTRIBUTING.md#conventional-commit-standard)
- [ ] tests and/or benchmarks are included
- [ ] documentation is changed or added
- [ ] commit message and code follows the [Code of conduct](https://github.com/wundergraph/wundergraph/blob/main/CODE_OF_CONDUCT.md)

### CODE REVIEW

File: packages/sdk/src/client/client.ts [hunk 1]
```diff
@@ -166,7 +166,7 @@ export class Client {
 			};
 		}
 
-		if (!resp.data) {
+		if (resp.data === undefined) {
```
Comment (Pagebakers) [inline packages/sdk/src/client/client.ts:169]: So null can be valid?

**Review by Pagebakers — APPROVED** [review]

File: packages/sdk/src/client/client.ts (hunk 1 above)
Comment (fiam) [reply packages/sdk/src/client/client.ts:169]: I think so, an operation might return null and we should handle that gracefully. Maybe I'm missing something? 😅 Keep in mind that we use this to handle return values from TS operations too, and they don't necessarily have to abide by GQL conventions.

File: pkg/apihandler/authentication.go [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+	}
+
+	if c := loadvariable.String(api.AuthenticationConfig.CookieBased.CsrfSecret); c != "" {
+		csrfSecret = []byte(c)
+	} else if fallback := api.CookieBasedSecrets.CsrfSecret; fallback != nil {
+		csrfSecret = fallback
+	}
+
+	if api == nil || api.HasCookieAuthEnabled() && (hashKey == nil || blockKey == nil || csrfSecret == nil) {
+		panic("API is nil or hashkey, blockkey, csrfsecret invalid: This should never have happened. Either validation didn't detect broken configuration, or someone broke the validation code")
```
Comment (StarpTech) [inline pkg/apihandler/authentication.go:46]: Why not return an error?

**Review by StarpTech — CHANGES_REQUESTED** [review]

File: pkg/apihandler/authentication.go (hunk 2 above)
Comment (fiam) [reply pkg/apihandler/authentication.go:46]: I'd say whoever added that block intended the API misuse to be very evident. You can ignore an error, but it's way more difficult to ignore a panic. I don't usually I don't think it's a bad idea.

Comment (StarpTech) [reply pkg/apihandler/authentication.go:46]: Panic should be only used when there is no other way. Here we can simply return an error and let the callee decide.

**Review by StarpTech — APPROVED** [review]
LGTM

**Comment by fiam** [comment]
Thanks @StarpTech!

**Pull request merged**